### DISCUSSION TITLE
Use basic auth or bearer token for securing endpoints.

### DISCUSSION THREAD

**Discussion by kklivil** [fastapi/fastapi#10982] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Code Snippet:
from fastapi.security import (
    HTTPBasic,
    HTTPBasicCredentials,
    HTTPBearer,
    HTTPAuthorizationCredentials,
)

basic_auth = HTTPBasic()
bearer_auth = HTTPBearer(auto_error=False)

async def get_authorization_header(
    request: Request,
    basic: HTTPBasicCredentials = Depends(basic_auth),
    bearer: HTTPAuthorizationCredentials = Depends(bearer_auth)
):
    # Extract auth header from request.
    auth_header = request.headers.get("Authorization")
    logging.info(f"Auth header: {auth_header}")
    if auth_header:
        return auth_header
    else:
        raise HTTPException(status_code=401, detail="Unauthorized")

def authenticate_via_sas(auth_header: str):

    sas_url = configs["sas"]["url"]
    headers = {"Authorization": auth_header}

    try:
        response = requests.get(sas_url, headers=headers)
        logging.info(f"Response: {response}")
    except requests.RequestException as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(e)
        )

    if response.status_code != status.HTTP_200_OK:
        logging.info(f"response: {response.status_code}")
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED, detail="Authentication failed"
        )

@app.get("/")
async def home_page(auth_header: str = Depends(get_authorization_header)):

    authenticate_via_sas(auth_header)

    with open("dist/index.html") as FH:
        return HTMLResponse(content=FH.read())


@app.post("/insights")
async def generate_insights(
    payload: Mia_Request, auth_header: str = Depends(get_authorization_header)
):
    authenticate_via_sas(auth_header)
    # rest of the code below ..
```

### Description

How to authorize end-points which can either receive **user name and password or Bearer token in the request** header. We have our simple authentication service which accepts the auth header and returns the Authorized or Not Authorized.

**Use case:** 
a. The browser should pop up a prompt for accepting user name and password e.g. `http://<ip_address>:8000` (endpoint "/")
b. If accessing endpoints via a client like the curl command or Postman the Fastapi should allow the request to reach get_authorization_header method if the request header contains a username and password or a bearer token.
c. The swagger page (/docs) should ask for a username and password or bearer token before allowing a user to access the endpoints. 

**Issue** Even though the bearer token is set in the request the FastAPI returns:
```
{
    "detail": "Not authenticated"
}
```
The control doesn't reach to the `get_authorization_header` method. 
Tested it using Postman.

### Operating System

Linux

### Operating System Details

`Ubuntu 20.04.6 LTS x86_64`

### FastAPI Version

0.95.1

### Pydantic Version

1.10.13

### Python Version

3.10

**Comment by jgould22** [comment, marked as answer]
For A and B  is expected behaviour if you look at the code here 
https://github.com/tiangolo/fastapi/blob/master/fastapi/security/http.py#L97

A couple things to think about for that
- In your impl you are not setting auto_error to false on your basic auth
- You probably need to set auto_error=False on basic auth and then handle the case where both deps return `None` in your `get_authorization_header` function
- Another option would be to write your own security dep and combine both your options (they are not especially complicated).

With regards to C you might find this helpful https://github.com/tiangolo/fastapi/discussions/8169 but I am not sure how your auth scheme might work with swagger/redoc
